_deploy/r/gnoswap/gns/gns.gno

```diff
@@ -93,93 +129,112 @@ func Render(path string) string {
+			mintUntilHeight = toHeight
 		}
-	}
 
-	err := privateLedger.Mint(users.Resolve(address), amountToMint)
-	if err != nil {
-		panic(err.Error())
+		// how many blocks to calculate
+		numBlock := uint64(mintUntilHeight-fromHeight) + 1
+
+		// amount of gns to mint for each block for current year
+		singleBlockAmount := GetAmountByHeight(yearEndHeight)
+
+		// amount of gns to mint for current year
+		yearAmountToMint := singleBlockAmount * numBlock
+
+		// if last block of halving year, handle left emission amount
+		if isLastBlockOfHalvingYear(mintUntilHeight) {
+			yearAmountToMint += handleLeftEmissionAmount(i, yearAmountToMint)
+			totalAmountToMint += yearAmountToMint
+		} else {
+			totalAmountToMint += yearAmountToMint
+		}
+		SetHalvingYearMintAmount(i, GetHalvingYearMintAmount(i)+yearAmountToMint)
+
+		// update fromHeight for next year (if necessary)
+		fromHeight = mintUntilHeight + 1
 	}
 
-	lastMintedHeight = nowHeight
+	return totalAmountToMint
+}
 
-	return amountToMint
+// isLastBlockOfHalvingYear returns true if the current block is the last block of a halving year.
+func isLastBlockOfHalvingYear(height int64) bool {
+	year := GetHalvingYearByHeight(height)
+	lastBlock := GetHalvingYearBlock(year)
+
+	return height == lastBlock
 }
 
-func Burn(from pusers.AddressOrName, amount uint64) {
-	owner.AssertCallerIsOwner()
-	fromAddr := users.Resolve(from)
-	checkErr(privateLedger.Burn(fromAddr, amount))
+// handleLeftEmissionAmount handles the left emission amount for a halving year.
+// It calculates the left emission amount by subtracting the halving year mint amount from the halving year amount.
+func handleLeftEmissionAmount(year int64, amount uint64) uint64 {
+	return GetHalvingYearAmount(year) - GetHalvingYearMintAmount(year) - amount
 }
 
-func checkAndHandleIfLastBlockOfHalvingYear(height int64, amount uint64) uint64 {
-	year := GetHalvingYearByHeight(height)
-	lastBlock := halvingYearBlock[year]
-	if height == lastBlock {
-		leftForThisYear := halvingYearAmount[year] - halvingYearMintAmount[year]
-		amount = leftForThisYear
-		return amount
+// skipIfSameHeight returns true if the current block height is the same as the last minted height.
+// This prevents multiple gns minting inside the same block.
+func skipIfSameHeight(lastMintedHeight, currentHeight int64) bool {
+	if lastMintedHeight == currentHeight {
+		return true
 	}
 
-	return amount
+	return false
 }
 
-func checkErr(err error) {
-	if err != nil {
-		panic(err.Error())
+// isEmissionEnded returns true if the emission is ended.
+// It returns false if the emission is not ended.
+func isEmissionEnded(year int64) bool {
+	if 1 <= year && year <= 12 {
+		return false
 	}
+
+	return true
 }
 
-// TODO:
-//  1. when emission contract mint gns reward, last executed height should be get from gns contract.
-//     mint function of gns contract and mintGns function of emission contract should be synchronized.
 func GetLastMintedHeight() int64 {
```

Comment:
We can add comments and categorize them

// setter

// getter

Reply:
b21feef0564837d3e80c6eebd5ae84fae9a958bc